pkg/controller/jaeger/jaeger_controller.go

```diff
@@ -195,3 +181,94 @@ func (r *ReconcileJaeger) handleDependencies(str strategy.S) error {
 
 	return nil
 }
+
+func (r *ReconcileJaeger) apply(jaeger v1alpha1.Jaeger, str strategy.S) (bool, error) {
+	if err := r.applyDeployments(jaeger, str.Deployments()); err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (r *ReconcileJaeger) applyDeployments(jaeger v1alpha1.Jaeger, desired []appsv1.Deployment) error {
+	opts := client.MatchingLabels(map[string]string{
+		"app.kubernetes.io/instance":   jaeger.Name,
+		"app.kubernetes.io/managed-by": "jaeger-operator",
+	})
+	depList := &appsv1.DeploymentList{}
+	if err := r.client.List(context.Background(), opts, depList); err != nil {
+		return err
+	}
+
+	// we now traverse the list, so that we end up with three lists:
+	// 1) deployments that are on both `desired` and `existing` (update)
+	// 2) deployments that are only on `desired` (create)
+	// 3) deployments that are only on `existing` (delete)
+	depInventory := inventory.ForDeployments(depList.Items, desired)
+	for _, d := range depInventory.Create {
+		log.WithFields(log.Fields{
+			"namespace":  jaeger.Namespace,
+			"instance":   jaeger.Name,
+			"deployment": d.Name,
+		}).Debug("creating deployment")
+		if err := r.client.Create(context.Background(), &d); err != nil {
+			return err
+		}
+	}
+
+	for _, d := range depInventory.Update {
+		log.WithFields(log.Fields{
+			"namespace":  jaeger.Namespace,
+			"instance":   jaeger.Name,
+			"deployment": d.Name,
+		}).Debug("updating deployment")
+		if err := r.client.Update(context.Background(), &d); err != nil {
+			return err
+		}
+	}
+
+	for _, d := range depInventory.Create {
+		if err := r.waitForStability(d); err != nil {
+			return err
+		}
+	}
+	for _, d := range depInventory.Update {
+		if err := r.waitForStability(d); err != nil {
```

Comment:
Just wondering what action should be taken if a problem occurs? Not deleting the existing deployment is definitely a step in the right direction, but do we need to revert any changes that have been done (or remove any new deployments)? May be this has to be handled in a separate PR as more advanced handling?

Reply:
Yeah, I think we need to think more about it during the advanced handling because it all depends on "when" it failed, as well as "where". For instance, if we successfully updated one deployment already, where do we get the previous definition for it, to roll it back?